File: sorting/MergeSort/code.py
def merge_sort(array):
    ##Stop when the len of the array is less or equal than one
    if len(array)>1:  
        #Calc the mid of the array
        mid = len(array) // 2 # // mean integer division

        #Create two arrays left and right
        lefthalf = array[:mid] 
        print(lefthalf)

        righthalf = array[mid:]
        print(righthalf)

        #Divide the subarrays left and right
        merge_sort(lefthalf)
        merge_sort(righthalf)
        
        ##I send the array as a argument to change the same array and not another
        merge(lefthalf, righthalf, array)


def merge(lefthalf, righthalf, array):
    index_array_left=0
    index_array_right=0
    k=0
    while index_array_left < len(lefthalf) and \
          index_array_right < len(righthalf):

        if lefthalf[index_array_left] < \
            righthalf[index_array_right]:
            
            #assign the less to the new array 
            array[k]=lefthalf[index_array_left]

            """As the less was an element in the lefthalf we dont 
            need to compare this again so we increase the index
            of the left array"""
            index_array_left=index_array_left+1
        
        else:
            array[k]=righthalf[index_array_right]

            """As the less was an element in the righthalf we dont 
            need to compare this again so we increase the index
            of the left array"""
            index_array_right=index_array_right+1

        #It is necesary increase the pos of the original array
        k=k+1

    ##add the remaining elements
    while index_array_left < len(lefthalf):
        array[k]=lefthalf[index_array_left]
        index_array_left=index_array_left+1
        k=k+1

    while index_array_right < len(righthalf):
        array[k]=righthalf[index_array_right]
        index_array_right=index_array_right+1
        k=k+1



array = [-10, 37, 98 , 0 ,12, 192, 5]
print("Original Array")
print(array)
merge_sort(array)

print("Sorted Array")
print(array)

File: Sorting/MergeSort/code.cpp
#include <stdio.h>
#include <iostream>
#include <vector>

using namespace std;

void show(int array[], int length_array);
void sort(int array[], int pos_ini, int pos_final);
void merge(int array[], int pos_ini, int pos_mid,int pos_final);



void show(int array [], int length_array){
    int index = 0;
    // cout<< "size : "<< array.size() <<endl;
    while(index<length_array){
        printf("%d ",array[index]);
        index = index +1;
    }
    printf("\n");
}

void sort(int array[], int pos_ini, int pos_final){

    /*In this condition the len of the array
      left and right half arrays will be
      of 1 element both */
    if(pos_final > pos_ini){
         //first calc the half point
        int pos_mid = (pos_ini+pos_final)/2;

        sort(array, pos_ini, pos_mid);
        sort(array, pos_mid +1, pos_final);
        merge(array, pos_ini, pos_mid, pos_final);

    }
}

void merge( int array[], int pos_ini, int pos_mid,int pos_final){


    int size_left = pos_mid - pos_ini + 1;
    int size_right =  pos_final - pos_mid;

    /* create temp arrays */
    int lefthalf[size_left], righthalf[size_right];

    for (int i = 0; i < size_left; i++)
        lefthalf[i] = array[pos_ini + i];
    for (int j = 0; j < size_right; j++)
        righthalf[j] = array[pos_mid + 1+ j];


    int index_right_half  = 0;
    int index_left_half = 0;
    int index = pos_ini;
    while (index_left_half < size_left && index_right_half < size_right) {

        if(lefthalf[index_left_half] <= righthalf[index_right_half]){
            array[index] = lefthalf[index_left_half];
            index_left_half = index_left_half +1;
        }else{
            array[index] = righthalf[index_right_half];
            index_right_half  = index_right_half+1;
        }
        index = index +1;
    }
    //Copy the remaining elements if there is any
    while( index_left_half < size_left){
        array[index] =  lefthalf[index_left_half];
        index_left_half = index_left_half +1;
        index = index +1;
    }
    while( index_right_half < size_right){
        array[index] =  righthalf[index_right_half];
        index_right_half = index_right_half +1;
        index = index +1;
    }
}

int main(){
    int array[] = {-10, 37, 98 , 0 ,12, 192, 5};
    int length_array = sizeof(array)/ sizeof(array[0]);
    show(array, length_array);
    sort(array, 0, length_array  -1);
    show(array, length_array);
}
